internal/inventory/gcpfetcher/fetcher_assets.go

```diff
@@ -0,0 +1,100 @@
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package gcpfetcher
+
+import (
+	"context"
+
+	"github.com/elastic/elastic-agent-libs/logp"
+
+	"github.com/elastic/cloudbeat/internal/inventory"
+	gcpinventory "github.com/elastic/cloudbeat/internal/resources/providers/gcplib/inventory"
+)
+
+type (
+	assetsInventory struct {
+		logger   *logp.Logger
+		provider inventoryProvider
+	}
+	inventoryProvider interface {
+		ListAllAssetTypesByName(ctx context.Context, assets []string) ([]*gcpinventory.ExtendedGcpAsset, error)
+	}
+	ResourcesClassification struct {
+		assetType      string
+		classification inventory.AssetClassification
+	}
+)
+
+var ResourcesToFetch = []ResourcesClassification{
+	{gcpinventory.CrmOrgAssetType, inventory.AssetClassificationGcpOrganization},
+	{gcpinventory.CrmFolderAssetType, inventory.AssetClassificationGcpFolder},
+	{gcpinventory.CrmProjectAssetType, inventory.AssetClassificationGcpProject},
+	{gcpinventory.ComputeInstanceAssetType, inventory.AssetClassificationGcpInstance},
+	{gcpinventory.ComputeFirewallAssetType, inventory.AssetClassificationGcpFirewall},
+	{gcpinventory.StorageBucketAssetType, inventory.AssetClassificationGcpBucket},
+	{gcpinventory.ComputeSubnetworkAssetType, inventory.AssetClassificationGcpSubnet},
+	{gcpinventory.IamServiceAccountAssetType, inventory.AssetClassificationGcpServiceAccount},
+	{gcpinventory.IamServiceAccountKeyAssetType, inventory.AssetClassificationGcpServiceAccountKey},
+}
+
+func newAssetsInventoryFetcher(logger *logp.Logger, provider inventoryProvider) inventory.AssetFetcher {
+	return &assetsInventory{
+		logger:   logger,
+		provider: provider,
+	}
+}
+
+func (f *assetsInventory) Fetch(ctx context.Context, assetChan chan<- inventory.AssetEvent) {
+	for _, r := range ResourcesToFetch {
+		f.fetch(ctx, assetChan, r.assetType, r.classification)
+	}
+}
+
+func (f *assetsInventory) fetch(ctx context.Context, assetChan chan<- inventory.AssetEvent, assetType string, classification inventory.AssetClassification) {
+	f.logger.Infof("Fetching %s", assetType)
+	defer f.logger.Infof("Fetching %s - Finished", assetType)
+
+	gcpAssets, err := f.provider.ListAllAssetTypesByName(ctx, []string{assetType})
+	if err != nil {
+		f.logger.Errorf("Could not fetch %s: %v", assetType, err)
+		return
+	}
+
+	for _, item := range gcpAssets {
+		assetChan <- inventory.NewAssetEvent(
```

Comment:
item can be one of many asset types, and depending on which type that is we'll want to use different conditional enrichers (like network, machine, instance, etc)

i plan to do this in a future PR

Reply:
Makes sense. This is what we do for most resources for now.